src/input.rs

```diff
@@ -135,233 +128,257 @@ impl State {
             InputEvent::Keyboard { event, .. } => {
                 let serial = SERIAL_COUNTER.next_serial();
                 let time = Event::time_msec(&event);
+                let pressed = event.state() == KeyState::Pressed;
 
-                let mut action = self.niri.seat.get_keyboard().unwrap().input(
+                let Some(Some(action)) = self.niri.seat.get_keyboard().unwrap().input(
                     self,
                     event.key_code(),
                     event.state(),
                     serial,
                     time,
-                    |self_, mods, keysym| {
-                        if event.state() == KeyState::Pressed {
-                            let config = self_.niri.config.borrow();
-                            action(&config, comp_mod, keysym, *mods).into()
-                        } else {
-                            FilterResult::Forward
+                    |this, mods, keysym| {
+                        let config = this.niri.config.borrow();
+                        let key_code = event.key_code();
+                        match (
+                            action(&config, comp_mod, keysym, *mods),
+                            this.niri.suppress_keys.entry(key_code),
+                        ) {
+                            (action, Entry::Occupied(mut entry)) => {
+                                let key = entry.get_mut();
+                                if pressed {
+                                    *key += 1;
+                                } else {
+                                    *key -= 1;
+                                }
+
+                                if *key == 0 {
+                                    entry.remove_entry();
+                                }
+
+                                FilterResult::Intercept(action)
+                            }
+                            // Only register suppresses on presses, since user can trigger release
+                            // due to matching action from non-action hotkey by releasing some
+                            // of the modifiers.
+                            (Some(action), Entry::Vacant(entry)) if pressed => {
+                                entry.insert(1);
+                                FilterResult::Intercept(Some(action))
+                            }
+                            (_, Entry::Vacant(_)) => FilterResult::Forward,
                         }
                     },
-                );
+                ) else {
+                    return;
+                };
 
-                // Filter actions when the session is locked.
-                if self.niri.is_locked() {
```

Comment:
I think if you change the `let Some(Some(` above to `let Some(mut action) =` then you can leave this part of the code completely unchanged, save for the `if !pressed { return; }`.

Reply:
Yeah, but why have the old logic with extra `None` and increasing indention for the giant `match`.